text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,338 @@
+
+    async connectedCallback() {
+        const { default: ctor } = await import('lightning/concreteComponent');
+        this.lazyConstructor = ctor;
+    }
+}
+```
+
+Renders:
+
+Before import completes:
+
+```html
+<div class="container"></div>
+```
+
+After import completes:
+
+```html
+<div class="container">
+    <lightning-concrete-component></lightning-concrete-component>
+</div>
+```
+
+## Motivation
+
+The main issue with the current implementation of the `lwc:dynamic` directive is that it does not respect the 1:1 mapping of the custom element name to constructor in the `CustomElementRegistry`.  This directly conflicts with the [custom elements specification](https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element) and will make adopting the native `CustomElementRegistry` more difficult in the future.
+
+For context, the current design requires a component author define an arbitrary custom element name which will be used as the name for all constructors passed to the `lwc:dynamic` directive, creating a 1:N mapping.
+
+For example:
+
+```html
+<template>
+  <x-lazy lwc:dynamic={customCtor}></x-lazy>
+</template>
+```
+
+```javascript
+import { LightningElement } from "lwc";
+
+export default class DynamicCtor extends LightningElement {
+  customCtor;
+
+  connectedCallback() {
+    this.loadCtor();
+  }
+
+  async loadCtor() {
+    const { default: ctor } = await import("c/customConstructor");
+    this.customCtor = ctor;
+  }
+}
+```
+
+Will always render the same tag regardless of the constructor.
+
+```html
+  <x-lazy>
+    <!-- c-custom-constructor content -->
+  </x-lazy>
+```
+
+This proposal aims to overcome the issues with the current design by exposing the custom element name at compile time and using it as the name of the dynamic component that is created at runtime.
+
+## Prior art
+
+Several other frameworks have the concept of dynamically rendering components:
+
+- [Svelte: `<svelte:component this={expression}>`](https://svelte.dev/docs#template-syntax-svelte-component)
+- [Vue: `<component :is="...">`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components)
+- [Angular: viewContainerRef](https://angular.io/guide/dynamic-component-loader)
+- [Lit: static expressions](https://lit.dev/docs/templates/expressions/#static-expressions)
+- [Aura: $A.createComponent](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/js_cb_dynamic_cmp_async.htm)
+
+Frameworks that use JSX such as [React](https://reactjs.org/docs/rendering-elements.html#rendering-an-element-into-the-dom) and [Stencil](https://stenciljs.com/docs/templating-jsx) can dynamically render components directly into the markup using expressions.
+
+The goal of managing dynamic components in LWC is to follow the same component lifecycle as all other LWC custom elements.  Having said that, taking a similar approach as Svelte and Vue seems the most natural and intuitive path forward.
+
+## Detailed design
```

Comment:
Add a note to mark this type of component as "risky"

Reply:
@caridy is the component being marked as "risky" to provide renderer hooks in compiled template?
I ask because we decided not to provide renderer hooks for custom elements because of how we cascade the renderer from owner to its children.

Reply:
@caridy it looks like custom renderer hooks currently [do not work with custom elements](https://github.com/salesforce/lwc/blob/master/packages/@lwc/template-compiler/src/shared/renderer-hooks.ts#L48-L54).  

Should I mention in this RFC that once custom renderer hooks work with custom elements they will be marked as "risky" or skip that for now?

Reply:
ah, right! you're right!